tests/system/model_monitoring/test_app.py

```diff
@@ -264,35 +301,60 @@ def _infer(
         ), "Outputs length does not match inputs"
 
     @classmethod
-    def _get_model_enpoint_id(cls) -> str:
+    def _get_model_endpoint_id(cls) -> str:
         endpoints = mlrun.get_run_db().list_model_endpoints(project=cls.project_name)
         assert endpoints and len(endpoints) == 1
         return endpoints[0].metadata.uid
 
-    def test_app_flow(self) -> None:
+    @pytest.mark.parametrize("with_training_set", [True, False])
+    def test_app_flow(self, with_training_set) -> None:
         self.project = typing.cast(mlrun.projects.MlrunProject, self.project)
-        self._log_model()
+        self._log_model(with_training_set)
+
+        for i in range(len(self.apps_data)):
+            if "with_training_set" in self.apps_data[i].kwargs:
+                self.apps_data[i].kwargs["with_training_set"] = with_training_set
+
+        # workaround for ML-5997
+        if not with_training_set:
+            self.apps_data.pop(0)
 
         with ThreadPoolExecutor() as executor:
-            executor.submit(self._submit_controller_and_deploy_writer)
+            executor.submit(
+                self._submit_controller_and_deploy_writer,
+                deploy_histogram_data_drift_app=with_training_set,  # workaround for ML-5997
+            )
             executor.submit(self._set_and_deploy_monitoring_apps)
-            future = executor.submit(self._deploy_model_serving)
+            future = executor.submit(self._deploy_model_serving, with_training_set)
 
         serving_fn = future.result()
 
         time.sleep(5)
-        self._infer(serving_fn)
+        self._infer(serving_fn, with_training_set=with_training_set)
         # mark the first window as "done" with another request
         time.sleep(
             self.app_interval_seconds
             + mlrun.mlconf.model_endpoint_monitoring.parquet_batching_timeout_secs
             + 2
         )
-        self._infer(serving_fn, num_events=1)
+        self._infer(serving_fn, num_events=1, with_training_set=with_training_set)
         # wait for the completed window to be processed
         time.sleep(1.2 * self.app_interval_seconds)
 
-        self._test_v3io_records(ep_id=self._get_model_enpoint_id())
+        dataset = load_iris()
+        inputs: set = (
+            {
+                mlrun.feature_store.api.norm_column_name(feature)
+                for feature in dataset.feature_names
+            }
+            if with_training_set
+            else {f"f{i}" for i in range(len(dataset.feature_names))}
+        )
+        outputs = {"p0"}
```

Comment:
1. Please move this to another method. It doesn't make sense to do it here, as it's the main flow, and this is some data prep.
2. I would avoid calling `load_iris` twice in the test.

Reply:
I moved it to  the _log_model method